### PR TITLE
ci-wrapper: prevent piping from disabling automatic colors

### PR DESCRIPTION
**Pull request by SkySkimmer**
Replace PR #18337

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot run full co

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: dev/ci/ci-wrapper.sh [hunk 1]
```diff
@@ -14,7 +17,16 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 cd "${DIR}/../.." || exit 1
 
 export TIMED=1
-bash "${DIR}/${CI_SCRIPT}" 2>&1 | tee "$CI_NAME.log"
+if { [ -t 1 ] || [ "$CI" ]; } && command -v script > /dev/null; then
+  # prevent piping from disabling auto colors / enable auto colors in CI
+  if ! [ "$TERM" = dumb ]; then export TERM=fake; fi
```
Comment (SkySkimmer) [inline dev/ci/ci-wrapper.sh:22]: the printenv above doesn't have TERM, but `set -x` says it evaluates to `dumb` here
WTF

Comment (SkySkimmer) [reply dev/ci/ci-wrapper.sh:22]: apparently bash sets `TERM` to `dumb` as a normal variable (not exported) if it's unset.

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
Now we get colors but some jobs get stuck in `git log -n 1`

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by ejgallego — APPROVED** [review]
:+1:

File: dev/ci/ci-wrapper.sh [hunk 2]
```diff
@@ -14,7 +15,31 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
+
+if [ "$color_wanted" ] && command -v script > /dev/null; then
+  # prevent piping from disabling auto colors / enable auto colors in CI
+    if [ "$CI" ]; then
+        export TERM=fake
+
+        # otherwise git log can get stuck waiting
+        # (but for some reason not always, eg https://gitlab.inria.fr/coq/coq/-/jobs/3638132
+        # vs failing jobs in that pipeline)
+        exec < /dev/null
```
Comment (SkySkimmer) [inline dev/ci/ci-wrapper.sh:34]: it still gets stuck in perennial, fiat crypto and fiat crypto legacy even with this

Why?
And why those jobs but not other WITH_SUBMODULES jobs?

Comment (JasonGross) [reply dev/ci/ci-wrapper.sh:34]: If it is `git log` that is stuck waiting, maybe you want `git --no-pager log`?

**Review by JasonGross — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Looks like it works https://gitlab.com/SkySkimmer/coq/-/pipelines/1096045874

File: dev/ci/ci-wrapper.sh [hunk 3]
```diff
@@ -14,7 +15,27 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
+if { [ -t 1 ] && ! [ "$TERM" = dumb ]; } || [ "$CI" ]
+then color_wanted=1
+else color_wanted=
+fi
+
+if [ "$color_wanted" ] && command -v script > /dev/null; then
+  # prevent piping from disabling auto colors / enable auto colors in CI
+    if [ "$CI" ]; then
+      export TERM=xterm-color
+      export GIT_PAGER=cat
```
Comment (silene) [inline dev/ci/ci-wrapper.sh:30]: Rather than forcing the output of all the git commands to go through `cat`, it might be better to just disable the pager.

```suggestion
      export GIT_PAGER=""
```

And here is a variant that does not depend on environment variables:

```suggestion
      git config --global core.pager ""
```

Comment (SkySkimmer) [reply dev/ci/ci-wrapper.sh:30]: We don't want to hit the global config when this runs on a developer's machine. Developers may want to run with the CI variable set for debugging weird errors.

Comment (silene) [reply dev/ci/ci-wrapper.sh:30]: Good point. Then just the empty pager.

**Comment by SkySkimmer** [comment]
@ejgallego ready to merge?

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
@coqbot: merge now

**Comment by herbelin** [comment]
I have incompatibilities on MacOS with the PR, where `script` is BSD-like. Option `--quiet` is `-q` and option `-c` is done by giving the command at the end. However, I did not find an equivalent to `script`'s option `--return`. Also, I could not make the variant `-F` of `--flush` working.

The following works: `script -q /dev/null bash "${DIR}/${CI_SCRIPT}" 2>&1 | tee "$CI_NAME.log` (though I don't see more colors, and I don't see where I'm supposed to see colors).

Alternatively, one can also radically disabling the change on MacOS by changing the test into `if [ "$color_wanted" ] && command -v script > /dev/null && ! [ "$OSTYPE" = darwin ]; then`.

**Comment by SkySkimmer** [comment]
>though I don't see more colors, and I don't see where I'm supposed to see colors

Are you running in emacs? What target did you try?

**Comment by herbelin** [comment]
Not in emacs. Under MacOs, any `make ci-target` fails with a syntax error: `script: illegal option -- -` (from `ci-wrapper.sh`).

**Comment by SkySkimmer** [comment]
But you said "The following works"?

**Comment by herbelin** [comment]
Indeed, I think I was not very clear, sorry.

With current master (after #18338), `make ci-any-target` fails on MacOS. If I change the `script` line with `script -q /dev/null bash "${DIR}/${CI_SCRIPT}" 2>&1 | tee "$CI_NAME.log`, then it does not fail.

Additionally, I was wondering what the `script` line is supposed to add. Assuming e.g. that an incantation of `script` compatible with both Linux and BSD is used, how can I observe that it does what is expected?

**Comment by SkySkimmer** [comment]
You can look for instance at https://gitlab.inria.fr/coq/coq/-/jobs/3698120 (ci-argosy)
Most notably `Warning` is yellow

**Comment by JasonGross** [comment]
> If I change the `script` line with `script -q /dev/null bash "${DIR}/${CI_SCRIPT}" 2>&1 | tee "$CI_NAME.log`, then it does not fail.

I think you need either `--return` / `-e`, or to do something like
```
rm -f "${CI_NAME}.ok"
script -q -c "bash '${DIR}/${CI_SCRIPT}' && touch ${CI_NAME}.ok" /dev/null 2>&1 | tee "${CI_NAME}.log"
rm "${CI_NAME}.ok
```
Notably:
```
$ script -q -c 'bash -c false; RET=$?; echo $RET; exit $RET'; echo $?
1
0
$ script -q -e -c 'bash -c false; RET=$?; echo $RET; exit $RET'; echo $?
1
1
```
and we want the overall script to pick up on the failure of children.

If you want a shorter test for the colors, by the way, consider:
- `echo foo | grep --color o` (color)
- `echo foo | grep --color o | tee foo` (no color)
- `bash -c 'echo foo | grep --color o'` (color)
- `script -q -c 'bash -c "echo foo | grep --color o"'` (color)
- `script -q -c 'bash -c "echo foo | grep --color o"' | tee foo` (still in color)

This is because script starts a terminal for its command.

I also think `-c` is similarly required on Linux, though I'm not sure, but `--flush` / `-f` is optional.  (Though it might be required to view logs in real-time, so it might be worth testing something like `script -h 2>/dev/null | grep -q -- --flush` to see if `--flush` is supported.)

**Comment by herbelin** [comment]
Ok, thanks.

Then, @SkySkimmer, I can confirm that changing the `script` line of `ci-wrapper.sh` with `script -q /dev/null bash "${DIR}/${CI_SCRIPT}" 2>&1 | tee "$CI_NAME.log"` works and show colors in MacOS.

@JasonGross, with BSD script, `script -q bash false; echo $?` (an equivalent to `script -q -c "bash -c false"; echo $?` afaiu) returns 1, which seems to indicate that BSD script returns the exit code of its command??

In any case, the following patch works for me (i.e. does not fail and shows color):
```
--- a/dev/ci/ci-wrapper.sh
+++ b/dev/ci/ci-wrapper.sh
@@ -29,7 +29,11 @@ if [ "$color_wanted" ] && command -v script > /dev/null; then
       export TERM=xterm-color
       export GIT_PAGER=
     fi
-  script --quiet --flush --return -c "bash '${DIR}/${CI_SCRIPT}'" /dev/null 2>&1 | tee "$CI_NAME.log"
+    if [ "$OSTYPE" = darwin ]; then
+        script -q /dev/null bash "${DIR}/${CI_SCRIPT}" 2>&1 | tee "$CI_NAME.log"
+    else
+        script --quiet --flush --return -c "bash '${DIR}/${CI_SCRIPT}'" /dev/null 2>&1 | tee "$CI_NAME.log"
+    fi
 else
   if [ "$color_wanted" ]; then
     >&2 echo 'script command not available, colors will be hidden'
```

**Comment by herbelin** [comment]
Do I make a PR with the proposed change?

**Comment by JasonGross** [comment]
This breaks the script that is supposed to print errors in the checks tab (cc @Zimmi48 ) and also broke the bug minimizer.  Can this be fixed?
